pkg/plugins/plugins.go

```diff
@@ -22,6 +29,113 @@ type PluginKey struct {
 	IsInternal     bool
 }
 
+// Implementaion stores implementation type (e.g. instance-storage) and its name (e.g. s3, mongo)
+type Implementaion struct {
+	Type string `json:"type"`
+	Name string `json:"implementation"`
+}
+
+// PluginMetadata about an installed plugin.
+type PluginMetadata struct {
+	Name            string          `json:"name"`
+	ClientPath      string          `json:"clientPath,omitempty"`
+	Implementations []Implementaion `json:"implementations"`
+	VersionInfo
+}
+
+// VersionInfo contains information from running the `version` command against the plugin.
+type VersionInfo struct {
+	Version string `json:"version"`
+	Commit  string `json:"commit"`
+	Author  string `json:"author,omitempty"`
+}
+
+type PluginProvider interface {
+	List() ([]string, error)
+	GetMetadata(string) (*PluginMetadata, error)
+}
+
+func NewFileSystem(config *config.Config) *fileSystem {
+	return &fileSystem{
+		Config: config,
+	}
+}
+
+type fileSystem struct {
+	*config.Config
+}
+
+func (fs *fileSystem) List() ([]string, error) {
+	pluginsDir, err := fs.GetPluginsDir()
+	if err != nil {
+		return nil, err
+	}
+
+	files, err := fs.FileSystem.ReadDir(pluginsDir)
+	if err != nil {
+		return nil, errors.Wrapf(err, "could not list the contents of the plugins directory %q", pluginsDir)
+	}
+
+	plugins := make([]string, 0, len(files))
+	for _, file := range files {
+		if !file.IsDir() {
+			plugins = append(plugins, file.Name())
+		}
+	}
+
+	return plugins, nil
+}
+func (fs *fileSystem) GetMetadata(pluginName string) (*PluginMetadata, error) {
+	r := NewRunner(pluginName)
+
+	// Copy the existing context and tweak to pipe the output differently
+	jsonB := &bytes.Buffer{}
+
+	mixinContext := *fs.Context
+	mixinContext.Out = jsonB
+	if !fs.Debug {
+		mixinContext.Err = ioutil.Discard
+	}
+	r.Context = &mixinContext
+
+	cmd := CommandOptions{Command: "version --output json"}
+	err := r.Run(cmd)
+	if err != nil {
+		return nil, err
+	}
+
+	var metadata PluginMetadata
+	err = json.Unmarshal(jsonB.Bytes(), &metadata)
+	if err != nil {
+		return nil, err
+	}
+
+	// make json.Marshal return `[]` instead of `nil` for not having implementations
+	if len(metadata.Implementations) == 0 {
+		metadata.Implementations = make([]Implementaion, 0)
+	}
+	return &metadata, nil
+}
+
+// PrintMetadata writes plugin metadata to `ctx.Out` as plaintext|json|yaml format based on `opts`
+func PrintMetadata(ctx *context.Context, opts version.Options, metadata PluginMetadata) error {
```

Comment:
I named method as `PrintMetadata` instead of `PrintVersion` as it doesn't only print version. If you agree with that we can change `mixins.PrintVersion` to `mixins.PrintMetadata` to make it consistent in other PR

Reply:
That's fine, the current method name made more sense when all it did was literally print "v0.1.0" 😀